File: Projeto JS/src/game/Tabuleiro.js
import { Peca } from './Peca.js';
import { Rei } from './Rei.js';
import { Rainha } from './Rainha.js';
import { Torre } from './Torre.js';
import { Bispo } from './Bispo.js';
import { Cavalo } from './Cavalo.js';
import { Peao } from './Peao.js';
import { Heroi } from './Heroi.js';
import { BoboDaCorte } from './BoboDaCorte.js';
import { Ladrao } from './Ladrao.js';
import { Templario } from './Templario.js';

export class Tabuleiro {
   
    constructor(popularTabuleiro = true) {
     
        this.tabuleiro = Array(8).fill(null).map(() => Array(8).fill(null));
        this.pecasNoTabuleiro = [];

        if (popularTabuleiro) {
            this.iniciarPecas();
        }
    }

    iniciarPecas() {
        const nomePecas = ["Torre", "Cavalo", "BoboDaCorte", "Principe", "Ladrao", "Bispo", "Templario"];
        
        // Embaralha as peças (equivalente a Collections.shuffle)
        for (let i = nomePecas.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [nomePecas[i], nomePecas[j]] = [nomePecas[j], nomePecas[i]];
        }

        const filadeTras = [];
        for (let i = 0; i < 6; i++) {
            filadeTras.push(nomePecas[i]);
        }
        filadeTras.push("Rei");
        filadeTras.push("Rainha");
        
        this.pecasNoTabuleiro = [...filadeTras, "Peao"];

      
        for (let i = filadeTras.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [filadeTras[i], filadeTras[j]] = [filadeTras[j], filadeTras[i]];
        }
        
        // Colocar peças pretas 
        for (let j = 0; j < 8; j++) {
            this.tabuleiro[0][j] = this.criarPeca(filadeTras[j], "preto");
            this.tabuleiro[1][j] = this.criarPeca("Peao", "preto");
        }
        
        // Colocar peças brancas 
        for (let j = 0; j < 8; j++) {
            this.tabuleiro[7][j] = this.criarPeca(filadeTras[j], "branco");
            this.tabuleiro[6][j] = this.criarPeca("Peao", "branco");
        }
    }

    getPecasNoTabuleiro() {
        return this.pecasNoTabuleiro;
    }

    getTabuleiro() {
        return this.tabuleiro;
    }

    getPeca(linha, coluna) {
        return this.tabuleiro[linha][coluna];
    }
    
    setPeca(linha, coluna, peca) {
        this.tabuleiro[linha][coluna] = peca;
    }

    moverPeca(linhaInicial, colunaInicial, linhaFinal, colunaFinal) {
        const peca = this.getPeca(linhaInicial, colunaInicial);
        if (peca === null) {
            console.log("Nenhuma peça encontrada na posição inicial.");
            return;
        }
        this.setPeca(linhaFinal, colunaFinal, peca);
        this.setPeca(linhaInicial, colunaInicial, null);
    }

    check(linha, coluna, corAtacante) {
        for(let i = 0; i < 8; i++) {
            for(let j = 0; j < 8; j++) {
                const peca = this.tabuleiro[i][j];
                if(peca !== null && peca.getCor() === corAtacante) {
                    if (peca instanceof Peao) {
                        const direcao = peca.getCor() === "branco" ? -1 : 1;
                        if (linha === i + direcao && Math.abs(coluna - j) === 1) {
                            return true;
                        }
                    } else {
                        if (peca.isMovimentoValido(this, i, j, linha, coluna)) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    encontrarRei(corDoRei) {
        for (let i = 0; i < 8; i++) {
            for (let j = 0; j < 8; j++) {
                const peca = this.getPeca(i, j);
                if (peca !== null && peca.getNomePeca() === "Rei" && peca.getCor() === corDoRei) {
                    return [i, j];
                }
            }
        }
        return null;
    }

    clonar() {
        const clone = new Tabuleiro(false);
        for (let i = 0; i < 8; i++) {
            for (let j = 0; j < 8; j++) {
                clone.setPeca(i, j, this.getPeca(i, j));
            }
        }
        return clone;
    }

    criarPeca(nome, cor) {
        console.log(`Tentando criar a peça: '${nome}' da cor '${cor}'`);
        switch (nome) {
            case "Rainha": return new Rainha(cor);
            case "Rei": return new Rei(cor);
            case "Torre": return new Torre(cor);
            case "Bispo": return new Bispo(cor);
            case "Cavalo": return new Cavalo(cor);
            case "Peao": return new Peao(cor);
            case "Principe": return new Heroi(cor);
            case "BoboDaCorte": return new BoboDaCorte(cor);
            case "Ladrao": return new Ladrao(cor);
            case "Templario": return new Templario(cor);
            default: 
            console.error(`--- ERRO FATAL: PEÇA NÃO ENCONTRADA ---`);
            console.error(`O nome '${nome}' não corresponde a nenhuma peça conhecida.`);
            return null;
        }
    }

    encontrarHeroi(corDoHeroi) {
    for (let i = 0; i < 8; i++) {
        for (let j = 0; j < 8; j++) {
            const peca = this.getPeca(i, j);
        
            if (peca !== null && peca.getNomePeca() === "Principe" && peca.getCor() === corDoHeroi) {
                return [i, j];
            }
        }
    }
    return null; 
}

setFuriaHeroi(linha, coluna, estaEmCheck) {
    const peca = this.getPeca(linha, coluna);
  
    if (peca instanceof Heroi) {
        peca.setReiEstaEmCheck(estaEmCheck);
    }
}

setModoDoBobo(linha, coluna, novoModo) {
    const peca = this.getPeca(linha, coluna);
   
    if (peca instanceof BoboDaCorte) {
        
        peca.setModo(novoModo);
    }
}
}

File: src/main/java/game/Tabuleiro.java
package game;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Tabuleiro {
    private Peca[][] tabuleiro = new Peca[8][8];
    List<String> pecasNoTabuleiro = new ArrayList<>();

    public List<String> getPecasNoTabuleiro() {
        return pecasNoTabuleiro;
    }

    public Tabuleiro() {

        // Usando a bibilioteca Collections, criamos um arrayList com o nome de
        // todas as peças, que podem ser usadas durante o jogo
        List<String> nomePecas = new ArrayList<>();
        nomePecas.add("Torre");
        nomePecas.add("Cavalo");
        nomePecas.add("BoboDaCorte");
        nomePecas.add("Principe");
        nomePecas.add("Ladrao");
        nomePecas.add("Bispo");
        nomePecas.add("Templario");

        // embaralha as peças (variação no setup inicial)
        Collections.shuffle(nomePecas);

        List<String> FiladeTras = new ArrayList<>();


        for (int i = 0; i < 6; i++) {
            FiladeTras.add(nomePecas.get(i));
        }
        FiladeTras.add("Rei");
        FiladeTras.add("Rainha");

        pecasNoTabuleiro = new ArrayList<>(FiladeTras);
        pecasNoTabuleiro.add("Peao");

        Collections.shuffle(FiladeTras);

        // Colocar peças pretas
        for (int j = 0; j < 8; j++) {
            Peca peca = criarPeca(FiladeTras.get(j), "preto");
            tabuleiro[0][j] = peca;
            Peca peao = criarPeca("Peao", "preto");
            tabuleiro[1][j] = peao;
        }

        // Colocar peças branca
        for (int j = 0; j < 8; j++) {
            Peca peca = criarPeca(FiladeTras.get(j), "branco");
            tabuleiro[7][j] = peca;
            Peca peao = criarPeca("Peao", "branco");
            tabuleiro[6][j] = peao;
        }
    }

    public boolean check(int linha, int coluna, String corAtacante, boolean checandoMovimentoDoRei) {
        for(int i = 0; i < 8; i++) {
            for(int j = 0; j < 8; j++) {
                Peca peca = tabuleiro[i][j];

                if(peca != null && peca.getCor().equals(corAtacante)) {
                    if (peca instanceof Peao) {
                        int direcao = peca.getCor().equals("branco") ? -1 : 1;


                        if (linha == i + direcao && Math.abs(coluna - j) == 1) {
                            return true; // É um ataque de peão válido.
                        }
                    } else if (peca instanceof BoboDaCorte) {
                        BoboDaCorte bobo = (BoboDaCorte) peca;
                        String modoOriginal = bobo.getModoAtual();
                        if (checandoMovimentoDoRei) {
                            List<String> listaDeModosBoboAtual = bobo.getListaModosBobo();

                            for (String modo : listaDeModosBoboAtual) {
                                bobo.setModo(modo); // Altera o modo temporariamente
                                // Verifica se o Bobo, neste modo, pode atacar o Rei
                                if (bobo.isMovimentoValido(this, i, j, linha, coluna)) {
                                    bobo.setModo(modoOriginal); // Restaura o modo original
                                    return true; // Rei está em cheque
                                }
                            }
                        } else {
                            if (bobo.isMovimentoValido(this, i, j, linha, coluna)) {
                                return true;
                            }
                        }
                    } else {

                        peca.isMovimentoValido(this, i, j, linha, coluna);
                        if (peca.isMovimentoValido(this, i, j, linha, coluna)) {
                            return true;
                        }
                    }
                }
            }
        }
        return false;
    }

    public int[] encontrarRei(String corDoRei) {
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                Peca peca = getPeca(i, j);
                if (peca != null && peca.getNomePeca().equals("Rei") && peca.getCor().equals(corDoRei)) {
                    return new int[]{i, j};
                }
            }
        }
        return null; // Não deveria acontecer em um jogo normal.
    }

    public int[] encontrarHeroi(String corDoHeroi) {
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                Peca peca = getPeca(i, j);
                if (peca != null && peca.getNomePeca().equals("Principe") && peca.getCor().equals(corDoHeroi)) {
                    return new int[]{i, j};
                }
            }
        }
        return null;
    }

    // Metodo para criar uma cópia exata do tabuleiro atual.
    public Tabuleiro clonar() {
        Tabuleiro clone = new Tabuleiro();
        Peca[][] novoTabuleiroArray = new Peca[8][8];
        for (int i = 0; i < 8; i++) {
            for (int j = 0; j < 8; j++) {
                // não modificoa  as peças em si, apenas suas posicoes.
                novoTabuleiroArray[i][j] = this.tabuleiro[i][j];
            }
        }
        clone.setTabuleiro(novoTabuleiroArray);
        return clone;
    }

    public Peca[][] getTabuleiro() {
        return tabuleiro;
    }

    public void setTabuleiro(Peca[][] tabuleiro) {
        this.tabuleiro = tabuleiro;
    }

    public Peca getPeca(int linha, int coluna) {
        return tabuleiro[linha][coluna];
    }

    public void setPeca(int linha, int coluna, Peca peca) {
        tabuleiro[linha][coluna] = peca;
    }

    public void setListaModosBobo(int linha, int coluna, List<String> listaAtualModos) {
        Peca peca = getPeca(linha, coluna);

        if (peca instanceof BoboDaCorte) {
            ((BoboDaCorte) peca).setListaModosBobo(listaAtualModos);
        }
    }

    public void setModoDoBobo(int linha, int coluna, String novoModo) {
        Peca peca = getPeca(linha, coluna);

        if (peca instanceof BoboDaCorte) {
            ((BoboDaCorte) peca).setModo(novoModo);
        }
    }

    public void setFuriaHeroi(int linha, int coluna, boolean estaEmCheck) {
        Peca peca = getPeca(linha, coluna);

        if (peca instanceof Heroi) {
            ((Heroi) peca).setReiEstaEmCheck(estaEmCheck);
        }
    }

    public Peca criarPeca(String nome, String cor) {
        switch(nome) {
            case "Rainha":
                return new Rainha(cor);
            case "Rei":
                return new Rei(cor);
            case "BoboDaCorte":
                return new BoboDaCorte(cor);
            case "Ladrao":
                return new Ladrao(cor);
            case "Bispo":
                return new Bispo(cor);
            case "Principe":
                return new Heroi(cor); // talvez você queira renomear Heroi -> Principe
            case "Cavalo":
                return new Cavalo(cor);
            case "Torre":
                return new Torre(cor);
            case "Templario":
                return new Templario(cor);
            case "Peao":
            default:
                return new Peao(cor);
        }
    }

    /**
     * Move uma peça de (linhaInicial, colunaInicial) para (linhaFinal, colunaFinal).
     */
    public void moverPeca(int linhaInicial, int colunaInicial, int linhaFinal, int colunaFinal) {
        Peca peca = getPeca(linhaInicial, colunaInicial);

        if (peca == null) {
            System.out.println("Nenhuma peça encontrada na posição inicial.");
            return;
        }

        setPeca(linhaFinal, colunaFinal, peca);   // coloca peça no destino
        setPeca(linhaInicial, colunaInicial, null); // limpa posição inicial

      /*  System.out.println("✅ " + peca.getNomePeca() + " (" + peca.getCor() + ") movida de " +
                "(" + linhaInicial + "," + colunaInicial + ") para (" + linhaFinal + "," + colunaFinal + ")"); */
    }
}
